book/src/development/emit_lints.md

```diff
@@ -0,0 +1,151 @@
+down a skeleton for this lint as the following:
+
+```rust
+impl<'tcx> LateLintPass<'tcx> for BarExpressions {
+    fn check_expr(&mut self, cx: &LateContext<'tcx>, expr: &'tcx Expr<'_>)  {}
+}
+```
+
+### Emitting Lint in `LateLintPass`
+
+Inside the trait method that we implement, we can write down the lint logic
+and the emit the lint with suggestions.
+
+Clippy's [diagnostics] provides quite a few diagnositc functions that we can
+use to emit lints. Take a look at the documentation to pick one that suits
+your lint's needs the best. Some common ones you will encounter in the Clippy
+repository includes:
+
+- [span_lint_and_help]: emits lint and provides a helpful message
+- [span_lint_and_sugg]: emits lint and provides a suggestion to fix the code
+
+```rust
+impl<'tcx> LateLintPass<'tcx> for BarExpressions {
+    fn check_expr(&mut self, cx: &LateContext<'tcx>, expr: &'tcx Expr<'_>)  {
+        // Imagine a `some_bar_expr_logic` that checks for requirements for emitting the lint
+        if some_bar_expr_logic(expr) {
+            span_lint_and_help(
+                cx,
+                FOO_FUNCTIONS,
+                expr.span,
+                "message on why the lint is emitted",
+                None,
+                "message that provides a helpful suggestion",
+            );
+        }
+    }
+}
+```
+
+> Note: According to [the rustc-dev-guide], the message should be matter of fact and avoid
+> capitalization and periods, unless multiple sentences are needed. When code or
+> an identifier must appear in a message or label, it should be surrounded with
+> single grave accents \`.
+
+## `EarlyLintPass`
+
+Emitting a lint with `EarlyLintPass` follows the same logic like `LateLintPass`.
+Take a look at the [EarlyLintPass][early_lint_pass] documentation, which
+provides an abundance of methods that we can implement for our lint.
+
+```rust
+pub trait EarlyLintPass: LintPass {
+    // Trait methods
+}
+```
+
+### Implement Trait Method in `EarlyLintPass`
+
+Similar to `LateLintPass`, we pick a trait method to implement if we choose
+`EarlyLintPass` for a specific lint we want to design. Using `FooFunctions`
+as an example, we will use [`check_fn`][early_check_fn] since it gives us
+access to various information about the function that is currently being checked.
+
+```rust
+impl EarlyLintPass for FooFunctions {
+    fn check_fn(&mut self, cx: &EarlyContext<'_>, fn_kind: FnKind<'_>, span: Span, _: NodeId) {
+        // TODO: Emit lint here
+    }
+}
+```
+
+### Emitting Lint in `EarlyLintPass`
+
+To emit the lint with `EarlyLintPass`, use `span_lint_and_help`
+to provide an extra help message. This is how it looks:
+
+```rust
+impl EarlyLintPass for FooFunctions {
+    fn check_fn(&mut self, cx: &EarlyContext<'_>, fn_kind: FnKind<'_>, span: Span, _: NodeId) {
+        // Imagine a `is_foo_function` that checks for functions named `foo`
+        if is_foo_function(fn_kind) {
```

Comment:
I wouldn't put examples in here, but rather explain lint passes and how they work in general. 

---

I think my issue with this file is, that it thows together lint emission and lint passes. I think both things deserve their own chapter

- Lint passes: What are those? How does rustc call them? How to implement them (high level)? Difference Early/Late?
- Lint emission: How to emit a lint in Clippy? How to get code snippets? How to produce suggestions? Applicability? Difference between suggestion/help/note?